docs/proposals/component-registration.md

```diff
@@ -0,0 +1,409 @@
+	Status ConditionStatus // True, False, Unknown
+	LastUpdateTime util.Time
+	LastTransitionTime util.Time
+	Reason string
+	Message string
+}
+```
+
+
+### Kubectl (CLI)
+
+Ready Service Endpoints can be queried with the existing `kubectl get endpoints <service-name>`.
+
+Add a `kubectl get externalserviceproviders` (`kubectl get esp` for short) command to list the external service provider instances, their address/port/name and their last known conditions. Components can be retrived with a component specific label, like `type=component` or `component=apiserver`.
+
+
+### Storage
+
+Store ExternalServiceProviders spec/status in etcd, with an apiserver storage wrapper.
+
+
+### Status Updates
+
+Add an external-service-provider-controller that regularly probes registered providers and updates their stored last known status.
+
+Add self-registration to the components that uses the new ExternalServiceProviders API (Create call) when the component starts.
+
+Add self-deregistration to the components that uses the new ExternalServiceProviders API (Remove call) when the component stops gracefully (e.g. after receiving SIGTERM).
+
+Add self-updating to the components that uses the new ExternalServiceProviders API (Update call) when the component errors/panics.
+
+Use a TCP probe (of `/healthz` or `/healthz/ping`) to check liveness, because it indicates a server is listening on the expected host/path, even if its responses are unhealthy.
+
+Update individual component `/healthz` endpoints with sub-resources to allow them to perform as an indicator of readiness, not just liveness. Readiness should include transitively checking the readiness of immediate dependencies (e.g. etcd or other databases).
+
+
+## Example
+
+The following terminal commands are an example of how the `kubectl get externalserviceproviders` command might work.
+
+TODO
+
+## Alternate Naming
+
+ExternalServiceProvider is a bit verbose for a top-level API resource name. Here are some other naming options for the same concept.
+
+ExternalTarget
+[External]Endpoint (singular)
+[External]ServiceTarget
+[External]ServiceInstance
+[External]ServiceTarget
+[External]Shell (Pod without peas)
+[External]ServiceIngress
+[External]ServiceUnit
+[External]ServiceMember
+
+We considered ExternalPod, but figured it would be too confusing, because it doesn't have to be implemented by a Pod.
+
+Including Endpoint in the name is also pretty confusing unless we deprecate and replace Endpoints over time.
+
+
+## Next Steps
+
+Since existing AddOns (e.g. kube-ui, kube-dns) already register themselves as internal services, this proposal puts them on a similar footing to externally hosted components.
+
+Services could be made to include both internal pods and external providers (in the same service).
+
+It was also suggested that this proposal overlaps somewhat with the Nodes API, and that nodes (kubelets/kube-proxy) could also be registered as ExternalServiceProviders.
```

Comment:
It wasn't really a well flushed out idea, just something we mentioned in the last google/mesosphere meeting. I can remove it.